### PR TITLE
Fixing the bug where map was not getting updated if the map wasn't present when the document was added using the add_documents call. 

### PR DESCRIPTION
**Pull request by adityabharadwaj198**
Fixing the bug where if the original document didn't contain any map fields, adding new maps with docs resulted in map type issues & later didn't allow us to properly update maps in subsequent  calls. 

Also added more robust testing where we are now testing the field_type metadata & running every test through 3 different docs

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
* Bug fix

* **What is the current behavior?** (You can also link to an open issue here)
* In case a user didn't send in a map in an add_docs request, subsequent addition of a map field in an update docs request will lead to the map not getting replaced 

* **What is the new behavior (if this is a feature change)?**
* Even if a user's original documents didn't have a map to begin with, updates to maps will work as expected. 

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No

* **Have unit tests been run against this PR?** (Has there also been any additional testing?)
* Added more unit tests. Now every unit test checks if the marqo_field_types metadata field is having the correct value for a field. This is checked after both updation and addition of docs 
* Manually tested the test case for which this was failing:
* Screenshot 2025-03-06 at 4 43 02 PM

* **Related Python client changes** (link commit/PR here)
* NA

* **Related documentation changes** (link commit/PR here)
* NA

* **Other information**:
* NA

* **Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: tests/integ_tests/tensor_search/integ_tests/test_add_documents_semi_structured.py [hunk 1]
```diff
@@ -781,4 +782,102 @@ def test_a_text_index_will_treat_a_url_as_text(self):
...
+                self.assertEqual(vespa_fields['marqo__field_types']['float_map.c'], MarqoFieldTypes.FLOAT_MAP.value, f"Expected float_map.c to have type {MarqoFieldTypes.FLOAT_MAP.value} for document {id}")
+                self.assertEqual(vespa_fields['marqo__field_types']['float_map.d'], MarqoFieldTypes.FLOAT_MAP.value, f"Expected float_map.d to have type {MarqoFieldTypes.FLOAT_MAP.value} for document {id}")
+                self.assertEqual(vespa_fields['marqo__field_types']['bool_field'], MarqoFieldTypes.BOOL.value, f"Expected bool_field to have type 'bool' for document {id}")
+                self.assertEqual(vespa_fields['marqo__field_types']['short_string_field'], MarqoFieldTypes.STRING.value, f"Expected short_string_field to have type 'string' for document {id}")
+                self.assertEqual(vespa_fields['marqo__field_types']['long_string_field'], MarqoFieldTypes.STRING.value, f"Expected long_string_field to have type 'string' for document {id}")
+                self.assertEqual(vespa_fields['marqo__field_types']['int_field'], MarqoFieldTypes.INT.value, f"Expected int_field to have type {MarqoFieldTypes.INT_MAP.value} for document {id}")
```
Comment (papa99do) [inline tests/integ_tests/tensor_search/integ_tests/test_add_documents_semi_structured.py]: nit. type in the assertion message does not match expected type.

File: tests/integ_tests/tensor_search/integ_tests/test_add_documents_semi_structured.py [hunk 2]
```diff
@@ -781,4 +782,102 @@ def test_a_text_index_will_treat_a_url_as_text(self):
...
+                self.assertEqual(vespa_fields['marqo__field_types']['float_field'], MarqoFieldTypes.FLOAT.value, f"Expected float_field to have type 'float' for document {id}")
+                self.assertEqual(vespa_fields['marqo__field_types']['bool_field2'], MarqoFieldTypes.BOOL.value, f"Expected bool_field2 to have type 'bool' for document {id}")
+                self.assertEqual(vespa_fields['marqo__field_types']['custom_vector_field'], MarqoFieldTypes.TENSOR.value, f"Expected custom_vector_field to have type 'custom_vector' for document {id}")
```
Comment (papa99do) [inline tests/integ_tests/tensor_search/integ_tests/test_add_documents_semi_structured.py]: Do we have a `custom_vector` type? (in assertion message) should it be `tensor`?

File: tests/integ_tests/tensor_search/integ_tests/test_add_documents_semi_structured.py [hunk 3]
```diff
@@ -781,4 +782,102 @@ def test_a_text_index_will_treat_a_url_as_text(self):
...
+                "content": "abcd",
+                "vector": [1.0] * 32
+            }
+        }
+        self.id_to_doc = {
+            '1': self.doc,
+            '2': self.doc2,
+            '3': self.doc3
+        }
+        self.add_documents(self.config, add_docs_params=AddDocsParams(
```
Comment (papa99do) [inline tests/integ_tests/tensor_search/integ_tests/test_add_documents_semi_structured.py:842]: Suggestion for an edge case:  add self.doc3 in another batch and do not specify tensor_field and multimodal_combo_field as tensor field. In this case. We expect both `tensor_field` and `tensor_subfield` to be just string. This verified that different batch can have different type for the same field.

File: tests/integ_tests/core/document/test_partial_update_semi_structured.py [hunk 4]
```diff
@@ -133,6 +136,11 @@ def test_partial_update_should_update_bool_field(self):
                 self.assertFalse(updated_doc['bool_field'], f"Expected bool_field to be False for document {id}")
                 self._assert_fields_unchanged(updated_doc, ['bool_field'])
 
+                # Verify field type
+                raw_vespa_doc = self.config.vespa_client.get_document(id, self.config.index_management.get_index(self.index.name).schema_name)
```
Comment (papa99do) [inline tests/integ_tests/core/document/test_partial_update_semi_structured.py]: we can use `self.index.schema_name`

File: tests/integ_tests/tensor_search/integ_tests/test_add_documents_semi_structured.py (hunk 1 above)
Comment (adityabharadwaj198) [reply tests/integ_tests/tensor_search/integ_tests/test_add_documents_semi_structured.py]: thanks for catching that!

File: tests/integ_tests/tensor_search/integ_tests/test_add_documents_semi_structured.py (hunk 2 above)
Comment (adityabharadwaj198) [reply tests/integ_tests/tensor_search/integ_tests/test_add_documents_semi_structured.py]: we don't have that. Thanks for catching that!

File: tests/integ_tests/core/document/test_partial_update_semi_structured.py (hunk 4 above)
Comment (papa99do) [reply tests/integ_tests/core/document/test_partial_update_semi_structured.py]: Also, this section of code seems to be duplicated a lot of times in this test, consider extract a method? something like `_assert_field_type(id, [(field_name, field_type)])`?

File: tests/integ_tests/core/document/test_partial_update_semi_structured.py [hunk 5]
```diff
@@ -156,400 +164,640 @@ def test_partial_update_should_update_int_field_to_int(self):
...
+                self.assertEqual(doc['int_map.d'], 2, f"Expected int_map.d to be 2 for document {id}")
+                
+                # Verify field type
+                raw_vespa_doc = self.config.vespa_client.get_document(id, self.config.index_management.get_index(self.index.name).schema_name)
+                vespa_fields = raw_vespa_doc.document.dict().get('fields')
+                self.assertEqual(vespa_fields['marqo__field_types']['int_map.d'], MarqoFieldTypes.INT_MAP.value, 
+                              f"Expected int_map.d to have type {MarqoFieldTypes.INT_MAP.value} for document {id}")
 
+                #Verify that the int_map.a and int_map.b is removed from the field types
+                self.assertEqual(vespa_fields['marqo__field_types'].get('int_map.a', None), None)
```
Comment (papa99do) [inline tests/integ_tests/core/document/test_partial_update_semi_structured.py]: nit. this is equivalent to `self.assertNotIn('int_map.a', vespa_fields['marqo__field_types'])`

File: tests/integ_tests/core/document/test_partial_update_semi_structured.py
Comment (papa99do) [inline tests/integ_tests/core/document/test_partial_update_semi_structured.py:1]: We have two test cases `test_partial_update_should_reject_updating_dict_to_int_field` and `test_updating_int_map_to_int` to cover updating numeric_map to int. Do we have any test case to cover updating numeric fields to numeric_map fields?

Also, can we write a test case just to cover adding a new map and then try to update it (with some fields removed)?

File: tests/integ_tests/tensor_search/integ_tests/test_add_documents_semi_structured.py (hunk 3 above)
Comment (adityabharadwaj198) [reply tests/integ_tests/tensor_search/integ_tests/test_add_documents_semi_structured.py:842]: thanks, i'll add that

File: tests/integ_tests/core/document/test_partial_update_semi_structured.py
Comment (adityabharadwaj198) [reply tests/integ_tests/core/document/test_partial_update_semi_structured.py:1]: good idea! I have added a test case to check this `test_updating_int_to_int_map`

File: tests/integ_tests/core/document/test_partial_update_semi_structured.py (hunk 5 above)
Comment (adityabharadwaj198) [reply tests/integ_tests/core/document/test_partial_update_semi_structured.py]: have created a method to do this as suggested by you above, so have removed this statement.

File: src/marqo/core/semi_structured_vespa_index/semi_structured_vespa_index.py [hunk 6]
```diff
@@ -538,8 +522,10 @@ def _handle_dict_field(
             # Set the appropriate field type based on the value type
             if isinstance(v, int):
                 field_types[f'{field_name}.{k}'] = MarqoFieldTypes.INT_MAP.value
+                field_types[f'{field_name}'] = MarqoFieldTypes.INT_MAP.value
```
Comment (papa99do) [inline src/marqo/core/semi_structured_vespa_index/semi_structured_vespa_index.py:525]: would this become a problem if the map contains both float and int? We probably should have a `numeric_map` field type for the map itself.
For example, in the original doc,  `'map1':  {'a': 1}`, and when we update it to `'map1': {'b': 2.0}`, its type stays at `INT_MAP`, which is not very accurate.

File: tests/integ_tests/core/document/test_partial_update_semi_structured.py [hunk 7]
```diff
@@ -112,6 +112,15 @@ def _assert_fields_unchanged(self, doc: Dict[str, Any], excluded_fields: List[st
             else:
                 self.assertEqual(value, doc.get(field, None), f'{field} is changed.')
 
+    def _assert_field_types(self, field_names, field_types, id):
+        raw_vespa_doc = self.config.vespa_client.get_document(id, self.index.schema_name)
+        vespa_fields = raw_vespa_doc.document.dict().get('fields')
+
+        for field_name, field_type in zip(field_names, field_types):
```
Comment (papa99do) [inline tests/integ_tests/core/document/test_partial_update_semi_structured.py]: nit. instead of zipping two lists, it might be more readable to do Tuple[(str, type)] .So when we pass them in, we can group the field name and type together. For example:
```
self._assert_field_types(id, [
  ('field1', BOOL), 
  ('field2', INT),
  ('field3', None)
])
```

File: src/marqo/core/semi_structured_vespa_index/semi_structured_vespa_index.py (hunk 6 above)
Comment (adityabharadwaj198) [reply src/marqo/core/semi_structured_vespa_index/semi_structured_vespa_index.py:525]: currently we don't support a map with both int & float keys within it. And we also don't support type changes of keys from float to int etc. If we want to have that in place it will require a bigger change I believe.

File: tests/integ_tests/core/document/test_partial_update_semi_structured.py (hunk 7 above)
Comment (adityabharadwaj198) [reply tests/integ_tests/core/document/test_partial_update_semi_structured.py]: done

File: src/marqo/core/semi_structured_vespa_index/semi_structured_vespa_index.py [hunk 8]
```diff
@@ -424,37 +417,31 @@ def _update_numeric_field(
...
-                
-                # Set appropriate field type based on numeric_type and whether it's a map
-                field_type = (MarqoFieldTypes.INT_MAP if numeric_type is int else MarqoFieldTypes.FLOAT_MAP) if is_map_type else (MarqoFieldTypes.INT if numeric_type is int else MarqoFieldTypes.FLOAT)
+            if not field_exists_in_original_doc or field_value_changed:
+                vespa_fields[vespa_field_name] = {"assign": field_value}
 
-                # Set field type metadata by creating a assigned statement
-                vespa_fields[vespa_field_types_field_name] = {"assign": field_type.value}
+                # Set field type metadata by creating an assign statement
+                vespa_fields[vespa_field_types_field_name] = {"assign": field_type}
```
Comment (papa99do) [inline src/marqo/core/semi_structured_vespa_index/semi_structured_vespa_index.py:436]: I noticed this change here. What is the type of `vespa_field_types`? is field_type a `str` or `MarqoFieldTypes`?

Comment (adityabharadwaj198) [reply src/marqo/core/semi_structured_vespa_index/semi_structured_vespa_index.py:436]: It's supposed to be str itself. In the [original code](https://github.com/marqo-ai/marqo/pull/1139/files/82a994ab1e96b09f7114901c0b52a788259b2848#diff-3d90f53e58a621604f7dcf97912190162a5239578b0ca47cbf9ac704536aa91dL447) you'll see that field_type was the MarqoFieldTypes enum, that's why we needed to do field_type.value. But in this PR I am populating it with the actual string value of that enum (https://github.com/marqo-ai/marqo/pull/1139/files/82a994ab1e96b09f7114901c0b52a788259b2848#diff-3d90f53e58a621604f7dcf97912190162a5239578b0ca47cbf9ac704536aa91dL447) so we don't need to do a .value again

**Review by papa99do — APPROVED** [review]
